Modernize Huawei LTE

scop opened the pull request:
## Breaking Change:

Configuration has been consolidated below `huawei_lte`.

Device tracker no longer uses `known_devices.yaml` but entity registry.

## Description:

Modernize Huawei LTE quite a bit, config entry support, unauthenticated access.

**Related issue (if applicable):** fixes #23819

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#10380

## Example entry for `configuration.yaml` (if applicable):
```yaml
huawei_lte:
  - url: http://192.168.100.1/
    username: admin
    password: something

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [ ] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

scop committed:
Modernization rework

- config entry support, with override support from huawei_lte platform in YAML
- device tracker entity registry support
- refactor for easier addition of more features
- internal code cleanups

MartinHjelmare commented on homeassistant/components/huawei_lte/__init__.py:
```diff
@@ -59,98 +118,231 @@
+    if sw_version:
+        device_data["sw_version"] = sw_version
+
+    device_registry = await dr.async_get_registry(hass)
```
Please limit this PR to just the required changes to implement config entries. We don't need to implement the device registry.

scop replied:
I believe doing this will make it possible to assign the hub and devices connected to it to an area. Why wouldn't that be desirable?

MartinHjelmare replied:
It's desirable, but we try to limit the scope of PRs. This is already a big PR. Everything we can move to later PRs is a good move.

scop replied:
Done.

MartinHjelmare commented on homeassistant/components/huawei_lte/__init__.py:
```diff
@@ -40,6 +59,34 @@
 
+DEVICE_TRACKER_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_MONITORED_CONDITIONS, default=DEVICE_TRACKER_KEYS): vol.All(
```
We don't allow monitored conditions in config anymore.
https://github.com/home-assistant/architecture/blob/master/adr/0003-monitor-condition-and-data-selectors.md

We can implement support for disabling entities instead.

https://developers.home-assistant.io/docs/en/entity_registry_disabled_by.html

scop replied:
ADR-0003 does not apply, it explicitly says "if that data is fetched in a single API request". The whole point here is that every category requires a separate API call/HTTP request.

MartinHjelmare replied:
Ok. Can we solve that problem by using the entity disable options instead?

scop replied:
I don't know. And wrt disabled by default, see my comment below.

MartinHjelmare commented on homeassistant/components/huawei_lte/config_flow.py:
```diff
@@ -0,0 +1,204 @@
+        except ResponseErrorException:
+            _LOGGER.warning("Response error", exc_info=True)
+            errors["base"] = "response_error"
+        except BaseException:
```
Don't except `BaseException`. If necessary except `Exception` for unknown errors and disable broad except pylint warning.

Preferably handle the connection error specifically.

scop replied:
Why would catching Exception and messing with pylint settings be preferable to just catching BaseException?

The API doesn't specify what the possible connection errors or exceptions are, so I don't think it's appropriate to dig out and use an implementation detail like that.

MartinHjelmare replied:
`BaseException` is broader in scope than `Exception`. We don't except or use `BaseException` in the source except in a couple of modules where a 3rd party library uses that.

scop replied:
I can see that, however it's pretty much a matter of taste and deserves documentation/reasoning beyond just "we don't do that" so that people don't need to waste time discussing it in reviews. For example in https://developers.home-assistant.io/docs/en/next/development_guidelines.html#use-new-style-string-formatting

Anyway, changed.

MartinHjelmare commented on homeassistant/components/huawei_lte/config_flow.py:
```diff
@@ -0,0 +1,204 @@
+
+        fields = [
+            (vol.Required(key, default=self.config_entry.options.get(key, True)), bool)
+            for key in chain(DEVICE_TRACKER_KEYS, SENSOR_KEYS, SWITCH_KEYS)
```
I suggest we just set a default disabled value for each entity type and not implement any integration options for this. The user can turn off new entities for this integration in the system options, and or disable individual entities via entity registry.

scop replied:
If I understand correctly, this suggestion again goes against the goal of being able to invoke a lot of different requests to the device, even ones that end up being used at all. Every request counts.

MartinHjelmare replied:
If the user disables an entity there won't be a request, I'm thinking.

scop replied:
Defaulting to disabled sounds to me as being against the spirit of ADR-0003 which kind of says everything should be enabled/exposed by default.

MartinHjelmare replied:
We have specifically added system level options to disable entities to allow the user to control this, so we don't need to implement the options on a integration level.

scop replied:
So, something like this?

1. At startup, fetch everything once. A fetch is necessary so we know what entities exist in the first place, and to get data for possible new component features that use data from categories that other existing features already use.
1. Create missing entities as default disabled.
1. Stop polling categories for which there are no enabled entities.
1. Start polling categories if one enables an entity that uses info from them.

The above would create a problem with the device tracker as its entities come and go; disabling its fetch doesn't work that well with the above basic rules. So it will probably need some more configuration in order to continue fetching even with all entities disabled, if one wants new ones to appear. ISTR there was something related already implemented in some other integrations, do you remember any?

MartinHjelmare commented on homeassistant/components/huawei_lte/device_tracker.py:
```diff
@@ -1,63 +1,52 @@
 
 @attr.s
-class HuaweiLteScanner(DeviceScanner):
+class HuaweiLteScanner:
```
We should implement the new `ScannerEntity` entity class.

https://github.com/home-assistant/home-assistant/blob/e0f1677296e6b7f06e393b4d341bc247de169d53/homeassistant/components/device_tracker/config_entry.py#L120

See unifi and zha device trackers for example.

scop replied:
Isn't that limited to/supposed to be only used with config entries? See comment below for not wanting to limit to those.

MartinHjelmare replied:
Yes. We have to use config entries. We'll import the config yaml into the config flow and get a config entry.

MartinHjelmare commented on homeassistant/components/huawei_lte/device_tracker.py:
```diff
@@ -1,63 +1,52 @@
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({vol.Optional(CONF_URL): cv.url})
 
-HOSTS_PATH = f"{KEY_WLAN_HOST_LIST}.Hosts.Host"
+def setup_scanner(hass, config, see, discovery_info=None):
```
This is the legacy interface. For config entries we need to use `async_setup_entry`.

scop replied:
I remember having a look at this, but IIRC didn't figure out how would I go about making the tracker available also via good old yaml support, not just config entries. The current implementation at least works with both.

MartinHjelmare replied:
We'll import the yaml into a config flow and save it as a config entry. This is the current standard to support both config yaml and config flow. There's no option to keep this legacy interface and start using config entries.

scop replied:
The components that do this that hit my quick grep seem to do it only once. I guess this means that after the import, further changes need to be made through the UI, and YAML changes no longer have any effect. Yet none of them document this behavior in the component docs. This UX doesn't sound good to me. Have I misunderstood something?

MartinHjelmare replied:
You're right that currently many integrations that import the yaml config into a config entry doesn't allow syncing a change to config yaml with the created config entry. We currently don't have a helper or recommendation here. But it's totally possible to do this if wanted.

Eg, checksum the imported yaml/dict and save that checksum in the config entry. On next start, calculate the checksum of the yaml config again and check if we need to update the config entry.

MartinHjelmare commented on homeassistant/components/huawei_lte/switch.py:
```diff
@@ -0,0 +1,73 @@
+"""Support for Huawei LTE switches."""
```
Please remove this platform for now. It should be a separate PR.

scop replied:
Done.

scop committed:
Remove log level dependent subscription/data debug hack

No longer needed, because pretty much all keys from supported
categories are exposed as sensors.

Closes https://github.com/home-assistant/home-assistant/issues/23819

scop committed:
Upgrade huawei-lte-api to 1.4.1

https://github.com/Salamek/huawei-lte-api/releases

scop committed:
Add support for access without username and password

scop commented:
So, here's a work in progress rework of the whole shebang, using entities disabled on discover, and entity enablement. Appears to work in principle, but I'm seeing some issues (see the "Work in progress" section).

scop commented:
Also, the passing of system_options to disable new entities when creating the config entry requires https://github.com/home-assistant/home-assistant/pull/27044 to be applied first.

balloob commented:
> If I go to entity registry and enable some disabled entities, after a while I receive these errors in the logs (and the whole component stops working). Dunno why this happens, or how to prevent it.

That happens when Home Assistant does the reload. Apparently your integration does not deal with reloads well. You need to make sure that when an entry is unloaded, you unload and detach all things that you set up. You should not return from that method until it is all set up. So don't create task but await the forward unload calls.

balloob commented:
So when you re-enable an entity, Home Assistant doesn't know where that entity is, nor can it instantiate it. So what happens is that we wait up to 30 seconds after the last change and then reload the integration.

https://github.com/home-assistant/home-assistant/blob/5c01dd483fc6bf2d15848b7f4ebcc99fb45bb705/homeassistant/config_entries.py#L823-L831

MartinHjelmare commented on tests/components/huawei_lte/test_config_flow.py:
```diff
@@ -0,0 +1,148 @@
+
+async def test_incorrect_credentials(hass, login_requests_mock):
+    """Test we show user form on invalid credentials."""
+    await _test_login_error(
```
It's possible to parametrize pytest tests:
https://docs.pytest.org/en/latest/parametrize.html#pytest-mark-parametrize

scop replied:
Done.

MartinHjelmare reviewed (commented):
Generally looks good I think. Beware of async vs sync context.

MartinHjelmare commented on homeassistant/components/huawei_lte/__init__.py:
```diff
@@ -60,97 +91,167 @@
+) -> bool:
+    """Unload config entry."""
+    router = hass.data[DOMAIN].routers.pop(config_entry.data[CONF_URL])
+    await hass.async_add_executor_job(router.cleanup)
```
Await the forward entry unload here.

scop replied:
Done.

chmielowiec committed:
Update huawei-lte-api to 1.4.3 (#27269)

scop committed:
Use subclass init instead of config_entries.HANDLERS

scop commented on homeassistant/components/huawei_lte/config_flow.py:
```diff
@@ -0,0 +1,197 @@
+
+
+@config_entries.HANDLERS.register(DOMAIN)
+class ConfigFlowHandler(config_entries.ConfigFlow):
```
Done.

scop commented:
Thanks a lot for the assistance, I'll address rest of the comments a bit later.

scop commented on homeassistant/components/huawei_lte/device_tracker.py:
```diff
@@ -1,63 +1,95 @@
+        if self._is_connected:
+            self._name = host.get("HostName", self.mac)
+            self._device_state_attributes = {
+                k: v for k, v in host.items() if k not in ("MacAddress", "HostName")
```
Do we have a helper method that transforms CamelCase to snake_case yet? That's what would be required here; the info comes (mostly) as CamelCase from the router API.

MartinHjelmare replied:
No, but some integrations already use the `stringcase` library.

scop replied:
Done; added some preprocessing to get better results, maybe we should extract that to a general purpose util function.

scop committed:
Convert device state attributes to snake_case

scop committed:
Simplify scanner entity initialization

scop committed:
Remove not needed hass reference from Router

scop committed:
Return explicit None from unsupported old device tracker setup

scop commented on homeassistant/components/huawei_lte/device_tracker.py:
```diff
@@ -1,63 +1,95 @@
+    _LOGGER.warning(
+        "Loading and configuring as a platform is no longer supported or "
+        "required, convert to enabling/disabling available entities"
+    )
```
Done.

MartinHjelmare replied:
`return None` was removed in a recent commit.

scop replied:
Yes, pylint wanted it gone.

MartinHjelmare replied:
It should `return None` explicitly.

https://github.com/home-assistant/home-assistant/blob/21ca936d33ca824614621fbe7a17c20c220b2291/homeassistant/components/device_tracker/setup.py#L64-L66

scop replied:
Matter of taste. Can't blame pylint, it really makes no practical difference if the `return None` is there or not, None is what the caller gets anyway, so it's fair in calling it useless. If I add it, I will also need to add a `# pylint: disable=useless-return` there. That would be the first occurrence of that disable in the whole tree, which hints that the code is leaning towards pylint's opinion on this at the moment.

But if this is a review blocker for you, I guess I can add it. Please confirm.

MartinHjelmare replied:
We should return explicitly since the caller handles the return value explicitly.

scop replied:
Done a few days ago.

scop commented on homeassistant/components/huawei_lte/device_tracker.py:
```diff
@@ -1,63 +1,95 @@
+    entities = []
+    for host in (x for x in hosts if x.get("MacAddress")):
+        entities.append(HuaweiLteScannerEntity(router, host["MacAddress"]))
+    async_add_entities(entities)
```
Done.

scop commented on homeassistant/components/huawei_lte/device_tracker.py:
```diff
@@ -1,63 +1,95 @@
+    _name: str = attr.ib(init=False, default="device")
+    _device_state_attributes: Dict[str, Any] = attr.ib(init=False, factory=dict)
+
+    def __attrs_post_init__(self):
```
Done.

scop commented on homeassistant/components/huawei_lte/__init__.py:
```diff
@@ -166,17 +267,107 @@ def _setup_lte(hass, lte_config) -> None:
+    async def async_added_to_hass(self) -> None:
+        """Connect to router update signal."""
+        self._disconnect_dispatcher = async_dispatcher_connect(
+            self.router.hass, UPDATE_SIGNAL, self._async_maybe_update
```
Done.

scop committed:
Mark unknown connection errors during config as such

scop committed:
Drop some dead config flow code

scop committed:
Run config flow sync I/O in executor

scop commented on homeassistant/components/huawei_lte/config_flow.py:
```diff
@@ -0,0 +1,197 @@
+        if DOMAIN in self.hass.data:
+            existing_urls.extend(  # yaml configs
+                url_normalize(x, default_scheme="http")
+                for x in self.hass.data[DOMAIN].routers
```
Good point, removed.

scop commented on homeassistant/components/huawei_lte/config_flow.py:
```diff
@@ -0,0 +1,197 @@
+                )
+            else:
+                try:
+                    conn = AuthorizedConnection(
```
Done.

scop commented on homeassistant/components/huawei_lte/config_flow.py:
```diff
@@ -0,0 +1,197 @@
+            _LOGGER.warning("Response error", exc_info=True)
+            errors["base"] = "response_error"
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.warning("Connection error", exc_info=True)
```
Done.

scop commented on homeassistant/components/huawei_lte/config_flow.py:
```diff
@@ -0,0 +1,197 @@
+        title = None
+        client = Client(conn)
+        try:
+            info = client.device.basic_information()
```
Moved to executor.

scop commented on homeassistant/components/huawei_lte/config_flow.py:
```diff
@@ -0,0 +1,197 @@
+        try:
+            info = client.device.basic_information()
+            title = info["devicename"]
+        except Exception:  # pylint: disable=broad-except
```
Whenever the api call fails for some reason, which can be due to a variety of documented or undocumented things. Also if the call succeeds but returned info doesn't contain `devicename`; moved that outside of the `try` for clarity.

MartinHjelmare replied:
If we can differentiate the exceptions we try to do that, ie we can catch more specific known exceptions first and end with a broad exception.

scop replied:
Given that what we do about the exception is the same in this particular case no matter what it is, and that action is just to debug log and ignore it, I don't think it's worth the trouble here.

scop committed:
Parametrize config flow login error tests

scop committed:
Forward entry unload to platforms

scop commented on homeassistant/components/huawei_lte/config_flow.py:
```diff
@@ -0,0 +1,197 @@
+            _LOGGER.warning("Connection error", exc_info=True)
+            errors[CONF_URL] = "connection_failed"
+        if errors:
+            logout()
```
Done.

scop committed:
Async/sync fixups

scop commented on homeassistant/components/huawei_lte/__init__.py:
```diff
@@ -60,97 +91,167 @@
+            new_options[CONF_RECIPIENT] = yaml_recipient
+        # Update entry if overrides were found
+        if new_data or new_options:
+            hass.config_entries.async_update_entry(
```
Should be taken care of now.

scop commented on homeassistant/components/huawei_lte/__init__.py:
```diff
@@ -166,17 +267,107 @@ def _setup_lte(hass, lte_config) -> None:
+
+    # Forward config entry setup to platforms
+    for domain in (DEVICE_TRACKER_DOMAIN, SENSOR_DOMAIN):
+        hass.async_create_task(
```
Done too.

scop commented:
Majority of comments addressed, but the entity enable/disable still doesn't quite work. It no longer raises an exception, but logs this:
```
2019-10-07 22:37:26 INFO (MainThread) [homeassistant.components.device_tracker] Setting up device_tracker.huawei_lte
2019-10-07 22:37:26 INFO (MainThread) [homeassistant.components.sensor] Setting up sensor.huawei_lte
2019-10-07 22:37:26 INFO (MainThread) [homeassistant.components.notify] Setting up notify.huawei_lte
```
...and after that pretty much everything related to the config entry stops working in the UI config, can no longer enable/disable anything even so that their state would follow the state just set etc (which did work before IIRC). Will have a look later.

scop committed:
Improve data subscription debug logging

scop committed:
Implement on the fly add of new and tracking of seen device tracker entities

MartinHjelmare commented on homeassistant/components/huawei_lte/device_tracker.py:
```diff
@@ -2,34 +2,95 @@
+    async_add_entities(new_entities, True)
+
+
+async def async_unload_entry(hass, config_entry):
```
Platforms don't support `async_unload_entry`. We have to unsubscribe from `async_unload_entry` in the component.

scop replied:
Oh, that's kind of surprising given that unloading needs to be forwarded to platforms. But handled in component now.

scop committed:
Handle device tracker entry unload cleanup in component

scop committed:
Remove unnecessary _async_setup_lte, just have code in async_setup_entry

scop committed:
Remove time tracker on unload

scop committed:
Fix to not use same mutable default subscription set for all routers

scop committed:
Merge branch 'dev' into huawei-lte-modernize

scop committed:
Pylint fixes

scop committed:
Remove some redundant defensive device tracker code

scop commented:
Ok, with this batch of commits, stuff appears to work fine. Reload woes were because I was inadvertently using a mutable default defaultdict for all router instances (filed a couple of related pylint RFE's right after that), and there were some other less significant ones here and there.

I believe this is pretty much in a shape for inclusion now, still pending #27044 or similar, and the `_better_snakecase` should probably be turned to a general utility functions so other things can take advantage of it.

scop committed:
Add back explicit get_scanner None return, hush pylint

scop committed:
Adjust approach to set system_options on entry create

scop commented:
Way to set `system_options` tweaked, and new hopefully less intrusive PR for it filed as #27612.

scop committed:
Enable some sensors on first add instead of disabling everything

scop committed:
Fix SMS notification recipients default value

scop commented:
Tweaked to enable a small known subset of sensors by default, no longer requires #27612.

scop committed:
Add option to skip new device tracker entities

scop committed:
Fix SMS notification recipient option default

scop committed:
Work around https://github.com/PyCQA/pylint/issues/3202

scop committed:
Merge branch 'dev' into huawei-lte-modernize

scop committed:
Merge branch 'dev' into huawei-lte-modernize

scop committed:
Remove unrelated type hint additions

scop commented:
Should be good to go now.

MartinHjelmare commented on homeassistant/components/huawei_lte/device_tracker.py:
```diff
@@ -1,63 +1,189 @@
+        await async_add_new_entities(hass, router.url, async_add_entities, tracked)
+
+
+async def async_add_new_entities(
```
This should be a callback instead of a coroutine since we don't await inside. Callbacks are a bit faster than coroutines.

scop replied:
You mean just removing the async (and renaming as appropriate), i.e. turning it to a regular function?

MartinHjelmare replied:
It's still good to name it `async_add_new_entities` so it's clear that it should run in the event loop. Just remove `async ` in `async def` and don't await it.

scop replied:
Azure confuses me. Tests failed with an apparently unrelated error (plex tests timeouting). When I click re-run, either nothing happens, or the test I tried to re-run gets silently immediately marked as cancelled without any explanation.

Oh well, maybe I'll just merge dev here again to get it to re-run.

MartinHjelmare commented on homeassistant/components/huawei_lte/config_flow.py:
```diff
@@ -0,0 +1,223 @@
+                (
+                    (
+                        vol.Optional(
+                            CONF_TRACK_NEW,
```
@balloob do we want each device tracker platform to implement this option?

It will be a lot of repetition, since probably all device scanners need this.

balloob replied:
We should remove this option. This is the same as the system option https://developers.home-assistant.io/docs/en/entity_registry_disabled_by.html#config-entry-system-options-setting-default-value-of-disabled_by-for-new-entity-registry-entries

scop replied:
The system option affects the entire router, whereas this option is just for the device tracker. Do you want it removed anyway?

scop replied:
Also, I would like to be able to not poll the associated device list if the user does not want anything from there. If I remove this option, do you think we should disable the poll if new entities are disabled?

Otherwise, we would have no option but to always enable the poll and accumulate disabled device tracker entities in the entity registry even if the user is never going to do anything with them, and the user could not do anything about that. That's something I'd like to avoid.

balloob replied:
The idea behind still filling up the entity registry is that a user can decide to enable specific entities that it wants to follow. It would be up to the integration to see if any entity needs updating (based on a callback to enable polling in `async_added_to_hass` callback) and only poll if necessary.

Since users will only be able to set this option after they create the integration, the sensors should already been added ?

scop committed:
Change async_add_new_entities to a regular function

scop committed:
Merge branch 'dev' into huawei-lte-modernize

scop commented:
Re better snakecase: https://github.com/okunishinishi/python-stringcase/pull/18

scop committed:
Remove option to disable polling for new device tracker entries

scop committed:
Merge branch 'dev' into huawei-lte-modernize

MartinHjelmare reviewed (approved):
Nice work!
